File: relay/src/wire/frame.test.ts
import { describe, expect, it } from 'vitest'
import {
  addressHint,
  build,
  controlFrame,
  decodeAddress,
  encodeAddress,
  FrameError,
  FrameType,
  HEADER_BYTES,
  parse,
  WIRE_VERSION,
} from './frame.js'

const address = (fill: number): Buffer => Buffer.alloc(16, fill)

const MAX = 256 * 1024

describe('the envelope', () => {
  it('comes back out exactly as it went in', () => {
    const raw = build({
      type: FrameType.SEALED,
      to: address(1),
      from: address(2),
      counter: 42n,
      body: Buffer.from('sealed bytes'),
    })

    const frame = parse(raw, MAX)

    expect(frame.version).toEqual(WIRE_VERSION)
    expect(frame.type).toEqual(FrameType.SEALED)
    expect(frame.to.equals(address(1))).toBe(true)
    expect(frame.from.equals(address(2))).toBe(true)
    expect(frame.counter).toEqual(42n)
    expect(frame.body.toString()).toEqual('sealed bytes')
  })

  /** The counter becomes part of a nonce in phase 3, so it has to survive the full 64-bit range. */
  it('carries a counter far past what fits in a number', () => {
    const counter = 2n ** 60n + 7n
    const raw = build({ type: FrameType.SEALED, to: address(1), from: address(2), counter, body: Buffer.alloc(0) })

    expect(parse(raw, MAX).counter).toEqual(counter)
  })

  it('refuses something shorter than a header', () => {
    expect(() => parse(Buffer.alloc(HEADER_BYTES - 1), MAX)).toThrow(FrameError)
  })

  it('refuses something over the size limit', () => {
    const raw = build({
      type: FrameType.SEALED,
      to: address(1),
      from: address(2),
      counter: 0n,
      body: Buffer.alloc(100),
    })

    expect(() => parse(raw, 50)).toThrow(FrameError)
  })

  it('refuses a version it does not know', () => {
    const raw = build({ type: FrameType.SEALED, to: address(1), from: address(2), counter: 0n, body: Buffer.alloc(0) })
    raw.writeUInt8(99, 0)

    expect(() => parse(raw, MAX)).toThrow(/wire version/)
  })

  it('refuses a frame type it does not know', () => {
    const raw = build({ type: FrameType.SEALED, to: address(1), from: address(2), counter: 0n, body: Buffer.alloc(0) })
    raw.writeUInt8(0x7f, 1)

    expect(() => parse(raw, MAX)).toThrow(/frame type/)
  })

  it('refuses an address that is not sixteen bytes', () => {
    expect(() =>
      build({ type: FrameType.SEALED, to: Buffer.alloc(8), from: address(2), counter: 0n, body: Buffer.alloc(0) }),
    ).toThrow(FrameError)

    expect(() => decodeAddress('too-short')).toThrow(FrameError)
  })

  it('writes an address as twenty-two characters and reads it back', () => {
    const text = encodeAddress(address(7))

    expect(text).toHaveLength(22)
    expect(decodeAddress(text).equals(address(7))).toBe(true)
  })

  /** A log line gets four bytes and never more: the address is the one thing this server does keep. */
  it('gives a log only the first four bytes of an address', () => {
    expect(addressHint(address(0xab))).toEqual('abababab')
  })

  /**
   * The relay's own note has a type of its own so that it can never be taken for something an agent
   * said - the receiver treats it as advice to resynchronise and never as content.
   */
  it('marks its own note as a different kind of frame', () => {
    const frame = parse(controlFrame(address(1), address(2)), MAX)

    expect(frame.type).toEqual(FrameType.CONTROL)
    expect(frame.body).toHaveLength(0)
  })
})

File: src/test/kotlin/io/github/crmapache/amazingclaudecode/remote/FrameTest.kt
package io.github.crmapache.amazingclaudecode.remote

import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

/**
 * The envelope has a twin on the relay's side (relay/src/wire/frame.ts), and the two are obliged to
 * agree byte for byte. What is checked here is the shape itself: the offsets, the limits, and the
 * refusals - the parts a mismatch would show up in first.
 */
class FrameTest {

    private fun address(fill: Byte) = ByteArray(Frame.ADDRESS_BYTES) { fill }

    private val max = 256 * 1024

    @Test
    fun `an envelope comes back out exactly as it went in`() {
        val raw = Frame.build(Frame.TYPE_SEALED, address(1), address(2), 42L, "sealed".toByteArray())

        val envelope = Frame.parse(raw, max)

        assertEquals(Frame.WIRE_VERSION, envelope.version)
        assertEquals(Frame.TYPE_SEALED, envelope.type)
        assertContentEquals(address(1), envelope.to)
        assertContentEquals(address(2), envelope.from)
        assertEquals(42L, envelope.counter)
        assertEquals("sealed", String(envelope.body))
    }

    @Test
    fun `the header is exactly forty-two bytes`() {
        val raw = Frame.build(Frame.TYPE_SEALED, address(1), address(2), 0L, ByteArray(0))

        assertEquals(Frame.HEADER_BYTES, raw.size)
    }

    /** The counter becomes part of a nonce in phase 3 - the whole range has to survive the trip. */
    @Test
    fun `a counter near the top of its range survives`() {
        val counter = Long.MAX_VALUE - 3
        val raw = Frame.build(Frame.TYPE_SEALED, address(1), address(2), counter, ByteArray(0))

        assertEquals(counter, Frame.parse(raw, max).counter)
    }

    @Test
    fun `something shorter than a header is refused`() {
        assertFailsWith<Frame.FrameException> { Frame.parse(ByteArray(Frame.HEADER_BYTES - 1), max) }
    }

    @Test
    fun `something over the size limit is refused`() {
        val raw = Frame.build(Frame.TYPE_SEALED, address(1), address(2), 0L, ByteArray(100))

        assertFailsWith<Frame.FrameException> { Frame.parse(raw, 50) }
    }

    @Test
    fun `a version we do not speak is refused`() {
        val raw = Frame.build(Frame.TYPE_SEALED, address(1), address(2), 0L, ByteArray(0))
        raw[0] = 99

        assertFailsWith<Frame.FrameException> { Frame.parse(raw, max) }
    }

    @Test
    fun `a frame type we do not know is refused`() {
        val raw = Frame.build(Frame.TYPE_SEALED, address(1), address(2), 0L, ByteArray(0))
        raw[1] = 0x7f

        assertFailsWith<Frame.FrameException> { Frame.parse(raw, max) }
    }

    @Test
    fun `an address is twenty-two characters and comes back whole`() {
        val text = Frame.encodeAddress(address(7))

        assertEquals(22, text.length)
        assertContentEquals(address(7), Frame.decodeAddress(text))
    }

    @Test
    fun `something that is not an address is refused`() {
        assertFailsWith<Frame.FrameException> { Frame.decodeAddress("short") }
        assertFailsWith<Frame.FrameException> { Frame.decodeAddress("!!! not base64 !!!") }
    }

    /** A log line gets four bytes and never more: the address is the one thing the relay does keep. */
    @Test
    fun `a log hint is four bytes of the address`() {
        val hint = Frame.addressHint(address(0xab.toByte()))

        assertEquals("abababab", hint)
        assertTrue(hint.length < Frame.encodeAddress(address(0xab.toByte())).length)
    }
}
